firestore-tos-extension/functions/src/index.ts

```diff
@@ -0,0 +1,239 @@
+    }
+
+    /** Set new acknowledgment */
+    const ack = {
+      ...tosDoc.data(),
+      acknowledged: data.acknowledged,
+      noticeType: data.noticeType,
+      ...acknowledgementDates,
+    };
+
+    /** Add Acknowledgment */
+    await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(data.tosId)
+      .withConverter(acknowledgementConverter)
+      .set(
+        { ...ack },
+        {
+          merge: true,
+        }
+      );
+
+    /** Find current acknowledgements */
+    const acknowledgements = await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .withConverter(acknowledgementConverter)
+      .get()
+      .then(($) => $.docs.map((doc) => doc.data()));
+
+    /** Set claims on user and return */
+    const claims = {};
+    claims[process.env.EXT_INSTANCE_ID] = acknowledgements;
+
+    const { customClaims: existingClaims } = await auth.getUser(
+      context.auth.uid
+    );
+
+    const updatedClaims = await auth.setCustomUserClaims(context.auth.uid, {
+      ...existingClaims,
+      ...claims,
+    });
+
+    /** send event if configured */
+    await eventChannel?.publish({
+      type: "firebase.google.v1.tos-accepted",
+      data: JSON.stringify(updatedClaims),
+    });
+  }
+);
+
+export const createTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      throw new functions.https.HttpsError(
+        "unauthenticated",
+        "No valid authentication token provided."
+      );
+    }
+
+    /** check if tos data has been provided  */
+    if (!data) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "No valid data provided."
+      );
+    }
+
+    /** check if tos data has been provided  */
+    if (!data.noticeType) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "Invalid notice type"
+      );
+    }
+
+    /** Set claims on user and return */
+    return db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos")
+      .doc(data.tosId)
+      .withConverter(termsConverter)
+      .set(
+        {
+          ...data,
+          creationDate: admin.firestore.FieldValue.serverTimestamp(),
+        },
+        { merge: true }
+      );
+  }
+);
+
+export const getTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      throw new functions.https.HttpsError(
+        "unauthenticated",
+        "No valid authentication token provided."
+      );
+    }
+
+    const { tosId, latest_only = false, custom_filter } = data;
+
+    const query = db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos");
+
+    if (custom_filter) {
```

Comment:
I was thinking we can remove this custom_filter concept, and just make notice_type a string. If you have a different kind of notice in your application, you don't really need an object to represent it, just a string would suffice.

Reply:
Notice type now modified to be an optional string